agent/utils/net/netutils_linux.go

```diff
@@ -0,0 +1,112 @@
+//go:build linux
+// +build linux
+
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package net
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net"
+
+	"slices"
+
+	"github.com/aws/amazon-ecs-agent/agent/utils/netlinkwrapper"
+	"github.com/vishvananda/netlink"
+)
+
+// Finds a link by provided mac address.
+func FindLinkByMac(netlinkClient netlinkwrapper.NetLink, mac string) (netlink.Link, error) {
+	parsedMac, err := net.ParseMAC(mac)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse mac '%s': %w", mac, err)
+	}
+
+	links, err := netlinkClient.LinkList()
+	if err != nil {
+		return nil, fmt.Errorf("failed to list all links: %w", err)
+	}
+
+	for _, link := range links {
+		attrs := link.Attrs()
+		if attrs != nil && bytes.Equal(parsedMac, attrs.HardwareAddr) {
+			return link, nil
+		}
+	}
+
+	return nil, errors.New("link was not found")
+}
+
+// Checks if the given link has a default route of the given IP family.
+//
+// Link provided can be `nil`. In that case this function will look at routes for
+// all links on the instance.
+func HasDefaultRoute(
+	netlinkClient netlinkwrapper.NetLink, link netlink.Link, ipFamily int,
+) (bool, error) {
+	routes, err := netlinkClient.RouteList(link, ipFamily)
+	if err != nil {
+		return false, fmt.Errorf("failed to list routes: %w", err)
+	}
+	if slices.ContainsFunc(routes, isDefaultRoute) {
+		return true, nil
+	}
+	return false, nil
+}
+
+// Checks if the provided route is a default route.
+//
+// A default route is defined here as a route with a Gateway and all IP addresses covered
+// in its destination.
```

Comment:
nit: imo it can be useful to add functionality comment here, 1) Checking for dst==nil 2) Checking 0.0.0.0/0 and ::/0

Reply:
Good point. Will do.

Reply:
Updated. I also added `ipFamily` parameter to this function so that the caller can specify which IP Family they are interested in. Updated the tests accordingly.